File: Top K Frequent Elements.py
#https://leetcode.com/problems/top-k-frequent-elements/description/?envType=problem-list-v2&envId=heap-priority-queue

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        hashmap={}
        for i in nums:
            if i not in hashmap:
                hashmap[i]=1
            else:
                hashmap[i]+=1
        result=[]
        maxi=-10**9+7
        element=0
        while len(result)!=k:
            for i in hashmap:
                if hashmap[i]>maxi:
                    maxi=hashmap[i]
                    element=i
            result.append(element)
            hashmap[element]=-10**9+7
            maxi=-10**9+7
        return result
            


        

File: Top K Frequent Elements.cpp
//https://leetcode.com/problems/top-k-frequent-elements/

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> map;
        for(int i:nums){
            map[i]++;
        }
        //reverse it
        unordered_map<int,vector<int>> map2;
        vector<int> frequency;
        for(auto &pair:map){
            map2[pair.second].push_back(pair.first);
            frequency.push_back(pair.second);
        }
        make_heap(frequency.begin(),frequency.end());
        vector<int> result;
        int size=0;
        int maxi;
        while (size!=k){
            maxi=frequency.front();
            result.push_back(map2[maxi].back());
            map2[maxi].pop_back();
            pop_heap(frequency.begin(),frequency.end());
            frequency.pop_back();
            size++;
        }
        return result;
    }
};
